packages/rln/src/contract/rln_base_contract.ts

```diff
@@ -33,31 +33,45 @@ export class RLNBaseContract {
    * Allows injecting a mocked contract for testing purposes.
    */
   public constructor(options: RLNContractInitOptions) {
-    // Initialize members and subscriptions
-    this.fetchMembers()
-      .then(() => {
-        this.subscribeToMembers();
-      })
-      .catch((error) => {
-        log.error("Failed to initialize members", { error });
-      });
-
     const {
       address,
       signer,
       rateLimit = DEFAULT_RATE_LIMIT,
       contract
     } = options;
 
-    this.validateRateLimit(rateLimit);
+    log.info("Initializing RLNBaseContract", { address, rateLimit });
 
     this.contract = contract || new ethers.Contract(address, RLN_ABI, signer);
     this.rateLimit = rateLimit;
 
-    // Initialize event filters
-    this._membersFilter = this.contract.filters.MembershipRegistered();
-    this._membershipErasedFilter = this.contract.filters.MembershipErased();
-    this._membersExpiredFilter = this.contract.filters.MembershipExpired();
+    try {
+      log.info("Setting up event filters");
+      // Initialize event filters
+      this._membersFilter = this.contract.filters.MembershipRegistered();
+      this._membershipErasedFilter = this.contract.filters.MembershipErased();
+      this._membersExpiredFilter = this.contract.filters.MembershipExpired();
+      log.info("Event filters initialized successfully");
+    } catch (error) {
+      log.error("Failed to initialize event filters", { error });
+      throw new Error(
+        "Failed to initialize event filters: " + (error as Error).message
+      );
+    }
+
+    // Initialize members and subscriptions
+    this.fetchMembers()
```

Comment:
I would recommend to improve it even further and:
- use similar approach to https://github.com/waku-org/js-rln/blob/f6d5deb8cb51602b2b544de5de5b88de96e11aef/src/contract/rln_contract.ts#L56 for both contract implementations;
- inject an instance of base contract into specific contract and not use `extend`

Reply:
> not use extend

What is being extended that you suggest changing?

Reply:
I ran into some membership cache management issues when I tried to compose it instead of inherit (seems like it needs some tests refactoring)

Creating a separate issue: https://github.com/waku-org/js-waku/issues/2392